src/combinations.rs

```diff
@@ -40,63 +43,118 @@ pub fn combinations<I>(iter: I, n: usize) -> Combinations<I>
+            }
+
+            if self.indices[0] == pool_len - self.n {
+                return None;
             }
 
             while self.indices[i] == i + pool_len - self.n {
-                if i > 0 {
-                    i -= 1;
-                } else {
-                    // Reached the last combination
-                    return None;
-                }
+                i -= 1;
             }
 
             // Increment index, and reset the ones to its right
             self.indices[i] += 1;
-            let mut j = i + 1;
-            while j < self.n {
+            for j in i + 1..self.n {
                 self.indices[j] = self.indices[j - 1] + 1;
-                j += 1;
             }
         }
 
+        self.pos += 1;
+
         // Create result vector based on the indices
         let mut result = Vec::with_capacity(self.n);
         for i in self.indices.iter() {
             result.push(self.pool[*i].clone());
         }
         Some(result)
     }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        if n == 0 {
+            return self.next();
+        } else if self.n == 0 {
+            return None;
+        }
+
+        // calculate 'absolute' position starting with 1
+        // to simplify calculations
+        let abs = n + self.pos + 1;
+        let mut pool_len = self.pool.len();
+        while pool_len - self.n + 1 < abs && self.pool.get_next() {
+            pool_len += 1;
+        }
+
+        let cnk = c_n_k(pool_len, self.n);
+        if self.pool.is_done() && cnk < abs {
+            self.pos = cnk;
+            return None;
+        }
+
+        // Here we are trying to calculate abs-th combination,
+        // counting from 1. We do so by calculating number of
+        // all possible combinations which have self.indices[ind] = pos.
+        let mut count = 0; // current count of already iterated combinations
+        let mut ind = 0; // current index (lies in 0..self.n)
+        let mut pos = 0; // current position in pool (lies in 0..pool_len)
+        while count <= abs && ind < self.n {
+            let cnk = c_n_k(pool_len - 1 - pos, self.n - ind - 1);
+            if count + cnk < abs {
+                // if less than abs combinations are possible,
+                // we should just try next position for current index
+                count += cnk;
+            } else {
+                // in this case, current position is 'saturated' and
+                // we need to start processing next index
+                self.indices[ind] = pos;
+                ind += 1;
+            }
+            pos += 1;
+        }
+
+        self.pos = abs;
+
+        // Create result vector based on the indices
+        let mut result = Vec::with_capacity(self.n);
+        for i in self.indices.iter() {
+            result.push(self.pool[*i].clone());
+        }
+        Some(result)
+    }
+}
+
+// C(n,k) is binomial coefficient.
+fn c_n_k(n: usize, k: usize) -> usize {
```

Comment:
You might want to consider another way to implement binomial coefficients, since this will overflow fairly quickly, for example when k is 21 or higher.

Some googling shows a fairly nice variant of binomial coefficients in num-integer: https://docs.rs/num-integer/0.1.39/src/num_integer/lib.rs.html#772-791

Reply:
Nice catch, thank you!
I was thinking how to do it better, the link you provided looks good.